FWCore/ParameterSet/python/SequenceTypes.py

```diff
@@ -830,13 +1228,192 @@ def __call__(self,test):
+
+    def _isTaskComponent(self):
+        return True
+
+    def isLeaf(self):
+        return False
+
+    def visit(self,visitor):
+        for i in self._order:
+            visitor.enter(i)
+            if not i.isLeaf():
+                i.visit(visitor)
+            visitor.leave(i)
+
+    def _errorstr(self):
+        return "Task(...)"
 
+    def __iter__(self):
+        for key in self._order:
+            yield key
+
+    def __str__(self):
+        l = []
+        v = ModuleNodeVisitor(l)
+        self.visit(v)
+        s = ''
+        for i in l:
+            if s:
+                s += ', '
+            s += str (i)
+        return s
+
+    def __repr__(self):
+        s = str(self)
+        return "cms."+type(self).__name__+'('+s+')\n'
+
+    def moduleNames(self):
+        """Returns a set containing the names of all modules being used"""
+        result = set()
+        visitor = NodeNameVisitor(result)
+        self.visit(visitor)
+        return result
+    def copy(self):
+        return Task(*self._order)
+    def copyAndExclude(self,listOfModulesToExclude):
+        """Returns a copy of the sequence which excludes those module in 'listOfModulesToExclude'"""
+        # You can exclude instances of these types EDProducer, EDFilter, ESSource, ESProducer,
+        # Service, or Task.
+        # Mostly this is very intuitive, but there are some complications in cases
+        # where objects that contain other objects are involved. See the comments
+        # for the _MutatingSequenceVisitor.
+        for i in listOfModulesToExclude:
+            if not i._isTaskComponent():
+                raise TypeError("copyAndExclude can only exclude objects that can be placed on a Task")
+        v = _CopyAndExcludeSequenceVisitor(listOfModulesToExclude)
+        self.visit(v)
+        return Task(*v.result(self))
+    def expandAndClone(self):
+        # Name of this function is not very good. It makes a shallow copy with all
+        # the subTasks flattened out (removed), but keeping all the
+        # modules that were in those subTasks as well as the top level
+        # ones.
+        l = []
+        v = ModuleNodeVisitor(l)
+        self.visit(v)
+        return Task(*l)
+    def replace(self, original, replacement):
+        """Finds all instances of 'original' and substitutes 'replacement' for them.
+           Returns 'True' if a replacement occurs."""
+        # This works for either argument being of type EDProducer, EDFilter, ESProducer,
+        # ESSource, Service, or Task.
+        #
+        # Mostly this is very intuitive, but there are some complications in cases
+        # where objects that contain other objects are involved. See the comments
+        # for the _MutatingSequenceVisitor.
+
+        if not original._isTaskComponent() or (not replacement is None and not replacement._isTaskComponent()):
+           raise TypeError("replace only works with objects that can be placed on a Task")
```

Comment:
Could be useful to also print out the type that was being used as a replacement.